src/promptflow/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -76,7 +95,28 @@ async def execute(
         # This is because it will always call `executor.shutdown()` when exiting the `with` block.
         # Then the event loop will wait for all tasks to be completed before raising the cancellation error.
         # See reference: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor
-        outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
+        try:
+            if line_timeout_sec is not None:
+                outputs = await self._execute(executor, nodes, inputs, context, line_timeout_sec)
+            else:
+                outputs = await self._execute_with_thread_pool(executor, nodes, inputs, context)
+        except Exception as e:
+            err_msg = "Flow execution has failed."
+            if isinstance(e, LineExecutionTimeoutError):
+                err_msg = f"Line execution timeout after {line_timeout_sec} seconds."
+                context.cancel_node_runs(err_msg)
+            node_names = ",".join(task.get_name() for task in self._running_tasks.values())
+            flow_logger.error(f"{err_msg} Cancelling all running nodes: {node_names}.")
+            current_running_task_names = list(self._running_tasks.keys())
```

Comment:
Seems no need to create a list of keys? It should be iterable.

Reply:
Offline discussed, it would cause dict size change during iteration, let's keep the current way.
Maybe use a more concise code, sth like:
```
for name in for name in current_running_task_names:
```
But it's optional

Reply:
In the same way, we change the `for task in running_tasks.values():` to `for task in list(running_tasks.values()):` to avoid the same problem.